src/coins.cpp

```diff
@@ -14,11 +14,107 @@
 #include <undo.h>
 #include <chainparams.h>
 
-/**
- * calculate number of bytes for the bitmask, and its number of non-zero bytes
- * each bit in the bitmask represents the availability of one output, but the
- * availabilities of the first two outputs are encoded separately
- */
+CCoins::CCoins() : fCoinBase(false), vout(0), nHeight(0), nVersion(0), originScId() { }
+
+CCoins::CCoins(const CTransactionBase &tx, int nHeightIn) {
+        FromTx(tx, nHeightIn);
+    }
+
+void CCoins::FromTx(const CTransactionBase &tx, int nHeightIn) {
+    fCoinBase  = tx.IsCoinBase();
+    vout       = tx.GetVout();
+    nHeight    = nHeightIn;
+    nVersion   = tx.nVersion;
+    originScId = tx.GetScId();
+    ClearUnspendable();
+}
+
+void CCoins::Clear() {
+    fCoinBase = false;
+    std::vector<CTxOut>().swap(vout);
+    nHeight = 0;
+    nVersion = 0;
+    originScId.SetNull();
+}
+
+void CCoins::Cleanup() {
+    while (vout.size() > 0 && vout.back().IsNull())
+        vout.pop_back();
+
+    if (vout.empty())
+        std::vector<CTxOut>().swap(vout);
+}
+
+void CCoins::ClearUnspendable() {
+    BOOST_FOREACH(CTxOut &txout, vout) {
+        if (txout.scriptPubKey.IsUnspendable())
+            txout.SetNull();
+    }
+    Cleanup();
+}
+
+void CCoins::swap(CCoins &to) {
+    std::swap(to.fCoinBase, fCoinBase);
+    to.vout.swap(vout);
+    std::swap(to.nHeight, nHeight);
+    std::swap(to.nVersion, nVersion);
+    std::swap(to.originScId, originScId);
+}
+
+bool operator==(const CCoins &a, const CCoins &b) {
+     // Empty CCoins objects are always equal.
+     if (a.IsPruned() && b.IsPruned())
+         return true;
+     return a.fCoinBase  == b.fCoinBase &&
+            a.nHeight    == b.nHeight   &&
+            a.nVersion   == b.nVersion  &&
+            a.vout       == b.vout      &&
+            a.originScId == b.originScId;
+}
+
+bool operator!=(const CCoins &a, const CCoins &b) {
+    return !(a == b);
+}
+
+bool CCoins::IsCoinBase() const {
+    return fCoinBase;
+}
+
+bool CCoins::IsFromCert() const {
+    // when restored from serialization, nVersion is populated only with latest 7 bits of the original value!
```

Comment:
this is true only with negative versions, pls update the comment

Reply:
Done